keps/sig-architecture/2019-07-16-node-role-label-use.md

```diff
@@ -0,0 +1,223 @@
+approved under. This was unintentional and due to unclear reviewer guidelines that have since been more strictly
+enforced. Likewise, the complexity of Kubernetes deployments has increased and the simplistic mapping of control
+plane concepts to a node has proven to limit the ability of conformant Kubernetes distributions to self-host, as
+anticipated.
+
+
+### Goals
+
+This KEP:
+
+* Clarifies that the use of the `node-role.kubernetes/*` label namespace is reserved solely for end-user and
+  external Kubernetes consumers, and:
+  * May not be used to vary behavior within Kubernetes projects that are subject to API review (kubernetes/kubernetes
+    and all components that expose APIs under the `*.k8s.io` namespace)
+  * Is not required to be used for a cluster to be conformant
+* Describes the locations within Kubernetes that must be changed to use an alternative mechanism for behavior
+  * Suggests approaches for each
+* Describes the timeframe and migration process for Kubernetes distributions and deployments to update labels
+
+
+## Proposal
+
+### Use of `node-role.kubernetes.io/*` labels
+
+* Kubernetes components MUST NOT set or alter behavior on any label within the `node-role.kubernetes.io/*` namespace.
+* Kubernetes components (such as `kubectl`) MAY simplify the display of `node-role.kubernetes.io/*` labels to convey the node roles of a node
+* Kubernetes examples and documentation MUST NOT leverage the node-role labels for node placement
+* External users, administrators, conformant Kubernetes distributions, and extensions MAY use `node-role.kubernetes.io/*` without reservation
+  * Extensions are recommended not to vary behavior based on node-role, but MAY do so as they wish
+* First party components like `kubeadm` MAY use node-roles to simplify their own deployment mechanisms.
+* Conformance tests MUST NOT depend on the node-role labels in any fashion
+
+
+### Migrating existing deployments
+
+The proposed fixes will all require deployment-level changes. That must be staged across several releases, and it should be possible for
+deployers to move early and "fix" the issues that may be caused by their topology.
+
+Therefore, for each change we recommend the following process to adopt the new labels in successive releases:
+
+* Release 1 (1.16):
+  * Introduce a feature gate for disabling node-role being honored. The gate defaults to on. `NodeRoleBehavior=on`
+  * Define the new node label with an associated feature gate for each feature area. The gate defaults to off. `FeatureA=off`
+  * Behavior for each functional area is defined as `NodeRoleBehavior == on && node_has_role || FeatureA == on && node_has_label`
+  * Deprecation of all node role behavior is announced
+  * No new components may leverage node-roles within Kubernetes projects.
+  * Early adopters may label their nodes to opt in to the features, even in the absence of the  gate.
```

Comment:
> even in the absence of the gate

it's a little unclear how that opts them in if there's also a per-feature gate that is off

Reply:
They would turn the gate on?

Reply:
They would turn the feature gate on and the legacy gate off.